### PR TITLE
[#155463709] Replace Spruce with Bosh CLI v2 features

### PR DESCRIPTION
**Pull request by camelpunch**
[#155463709 Migrate Spruce `grab` to bosh variables](https://www.pivotaltracker.com/story/show/155463709)

## What

We migrate our manifests to be based on [cf-deployment](https://github.com/cloudfoundry/cf-deployment). 

As a first step, we want to replace the spruce interpolation of basic values to bosh v2 variable interpolation:

 - Remove any "spruce" operators
 - Only use spruce for merge of big chunks of .yaml (aka stubs), but not basic values.

To implement this, we were forced to do a big refactoring to "flatten" the variables used, because bosh does not support deep merge of values in the YAML variable files. We renamed all variables so they only have one root key. For instance `secrets.uaa_admin_password` is renamed to `secrets_uaa_admin_password`. 

Check commit history for details.

## Dependencies

 - https://github.com/alphagov/paas-team-manual/pull/188

## How to review

Our strategy is to do a **No-op replacement** so that all the changes result in exactly the same manifest.

To test this:

 1. run the pipeline from master with Datadog enabled:

```
export DEPLOY_ENV=hector
git checkout master

make dev pipelines BRANCH=$(git rev-parse --abbrev-ref HEAD)  SELF_UPDATE_PIPELINE=false DEPLOY_ENV=hector ENABLE_DATADOG=true
# Run pipeline
```

 2. Make a copy the generated manifests 

```
aws s3 cp s3://gds-paas-${DEPLOY_ENV}-state/cf-manifest.yml  cf-manifest-master.yml 
aws s3 cp cf-manifest-master.yml  s3://gds-paas-${DEPLOY_ENV}-state/
aws s3 cp s3://gds-paas-${DEPLOY_ENV}-state/cloud-config.yml cloud-config-master.yml
aws s3 cp  cloud-config-master.yml s3://gds-paas-${DEPLOY_ENV}-state/
```
Ensure that the manifest has no changes, and that running this through a dev env with DataDog enabled has no effect.

 3. run the pipeline from the branch with Datadog enabled:

```
export DEPLOY_ENV=hector
git checkout bosh_interpolate_manifest_155463709

make dev pipelines BRANCH=$(git rev-parse --abbrev-ref HEAD)  SELF_UPDATE_PIPELINE=false DEPLOY_ENV=hector ENABLE_DATADOG=true
# Run pipeline
```
 4. Diff the manifests
```
diff --color -Nur <(aws s3 cp s3://gds-paas-${DEPLOY_ENV}-state/cf-manifest-master.yml - )  <(aws s3 cp s3://gds-paas-${DEPLOY_ENV}-state/cf-manifest.yml - ) 
diff --color -Nur <(aws s3 cp s3://gds-paas-${DEPLOY_ENV}-state/cloud-config-master.yml - )  <(aws s3 cp s3://gds-paas-${DEPLOY_ENV}-state/cloud-config.yml - ) 

```
### Other things to test

To test the secrets rotation, you can do:

```
./manifests/cf-manifest/scripts/generate-cf-secrets.rb > test.yml
./manifests/cf-manifest/scripts/rotate-cf-secrets.rb < test.yml > test_rotated.yml
diff --color -Nur test.yml test_rotated.yml
```

## Who can review

Not @camelpunch or @keymon or @LeePorte

### CODE REVIEW

File: concourse/pipelines/create-cloudfoundry.yml [hunk 1]
```diff
@@ -1122,8 +1122,8 @@ jobs:
             image_resource:
               type: docker-image
               source:
-                repository: governmentpaas/spruce
-                tag: 465642da06051a55630d39c899697b678f66a7f7
+                repository: keymon/bosh-cli-v2
```
Comment (dcarley) [inline concourse/pipelines/create-cloudfoundry.yml]: Could you rebase these container changes (and one below) out? I found them confusing when reviewing the individual commits and they might confuse somebody looking at the history.

File: manifests/cf-manifest/cloud-config/020-cf-vm-types.yml [hunk 2]
```diff
@@ -1,13 +1,10 @@
...
-    bosh:
-      password: (( grab secrets.vcap_password ))
-
 vm_types:
   - name: small
     network: cf
-    env: (( grab meta.default_env ))
+    env:
+      bosh:
+        password: ((secrets_vcap_password))
```
[the remark this answers is not in the document]
Comment (keymon) [reply manifests/cf-manifest/cloud-config/020-cf-vm-types.yml]: I would rather create a follow up story.

File: manifests/cf-manifest/manifest/data/000-aws-rds-combined-ca-bundle-pem.yml [hunk 3]
```diff
@@ -1,291 +1,286 @@
-meta:
-  ca_certs:
-    # AWS RDS bundle with all the root and intermediate certificates
-    # for all regions. More info:
-    # - http://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.SSL.html
-    # - http://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/CHAP_PostgreSQL.html#PostgreSQL.Concepts.General.SSL
```
[the remark this answers is not in the document]
Comment (keymon) [reply manifests/cf-manifest/manifest/data/000-aws-rds-combined-ca-bundle-pem.yml]: :+1:

File: manifests/cf-manifest/manifest/800-logsearch.yml [hunk 4]
```diff
@@ -162,17 +162,17 @@ jobs:
...
-      host: (( grab jobs.queue.networks.cf.static_ips.[1] ))
+      host: ((queue_static_ips_second))
   update:
     serial: true
 
 properties:
   syslog_daemon_config:
     enable: false
-  parser_logstash:
+  parser_logstash: # FIXME: is this still used?
```
[the remark this answers is not in the document]
Comment (keymon) [reply manifests/cf-manifest/manifest/800-logsearch.yml]: ok, removed.

File: manifests/variables.yml [hunk 5]
```diff
@@ -0,0 +1,6 @@
+---
+cell_instance_type: r4.xlarge
+cell_instance_memory_mb: 30662
+cell_rep_commit_percentage: 1.66
+# FIXME: This value is mb*percent
```
[the remark this answers is not in the document]
Comment (keymon) [reply manifests/variables.yml]: OK, removed and added a commend.

**Comment by keymon** [comment]
Regarding the keymon/bosh-cli-v2 container

> Could you rebase these container changes (and one below) out? I found them confusing when reviewing the individual commits and they might confuse somebody looking at the history.

It is a nightmare of rebase. do you really require it?

I can instead drop the container change and add a one-liner that downloads spruce, to keep the difference clear.

**Comment by dcarley** [comment]
Thanks!

File: .travis.yml [hunk 6]
```diff
@@ -47,7 +54,7 @@ before_install:
   - rsync -az ${TRAVIS_BUILD_DIR}/ $HOME/gopath/src/github.com/alphagov/paas-cf/
   - export TRAVIS_BUILD_DIR=$HOME/gopath/src/github.com/alphagov/paas-cf
   - cd $HOME/gopath/src/github.com/alphagov/paas-cf
-  - cd scripts && BUNDLE_GEMFILE=Gemfile bundle install --jobs=3 --retry=3 --deployment && cd ..
+  - bundle install --jobs=3 --retry=3 --deployment
```
Comment (alext) [inline .travis.yml]: This is unnecessary because Travis already takes care of installing the top-level Gemfile.

File: Gemfile [hunk 7]
```diff
@@ -1,9 +1,11 @@
 source "https://rubygems.org"
 
-gem 'bosh_cli', '1.3262.26.0'
+gem 'aws-sdk', '~> 2.3.8'
+gem 'cf-uaa-lib', '~> 3.8.0'
```
Comment (alext) [inline Gemfile]: This failed to add cf-uaac that was in the previous Gemfile. This is needed for interacting with UAA (as documented here: https://government-paas-team-manual.readthedocs.io/en/latest/guides/CF_UAA_user_management/)

**Pull request merged**